File: partition.js
var partition = function(head, x) {
let leftLinkedList = new ListNode(),rightLinkedList = new ListNode();
//make two lists denoting left and right parts
let it = head;//iterator
head = leftLinkedList;//to mark starting of left part (new head)
let temp = rightLinkedList;//to mark starting on right part
while(it != null){
    if(it.val<x){//node will be in left part
        leftLinkedList.next = it;
        leftLinkedList = leftLinkedList.next;
    }
    else{//node will be in right part
        rightLinkedList.next = it;
        rightLinkedList = rightLinkedList.next;
    }
    it = it.next;
    
}
rightLinkedList.next = null;//last node of right part will be last node of list
leftLinkedList.next = temp.next;//connect two lists

return head.next;
};

File: partition.java
/**
Input: head = [1,4,3,2,5,2], x = 3
Output: [1,2,2,4,3,5]
 */
class Solution {
    public ListNode partition(ListNode head, int x) {
        ListNode left = new ListNode(0);
        ListNode right = new ListNode(0);
        
        ListNode leftTail = left;
        ListNode rightTail = right;
        
        while(head != null){
            if(head.val < x){
                leftTail.next = head;
                leftTail = leftTail.next;
            }
            else{
                rightTail.next = head;
                rightTail = rightTail.next;
            }
            head = head.next;
        }
        
        leftTail.next = right.next;
        rightTail.next = null;
        
        return left.next;
    }
}
